changes to server ACL and subnets support

vchlum opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->

1. I would like to introduce subnets to `acl_hosts` and forward hostname resolution for its evaluation. Currently, the implementation of acl hosts is not suitable as a firewall substitution. This new feature (due to the forward hostname resolution) makes it a bit safer and moves it a bit closer to it. This means e.g.: `acl_hosts = 192.168.0.0/255.255.0.0` or   `acl_hosts = 192.168.0.0/16` enables access to subnet 192.168.?.?.

2. Enabling `gss`/Kerberos and `acl_hosts`, the `acl_hosts` is evaluated for `PBS_BATCH_Connect` even if the connection is authenticated by `gss`.  Using gss,  `acl_hosts` is checked to hostname while the request is `PBS_BATCH_Connect` and for others requests the `acl_krb_realms` is used to check the principals. It means both acls are needlessly used for the same connection.

3. Currently admin access can be obtained only via `resvport` auth method. I would like to introduce a server env `PBS_PRIVILEGED_AUTH_USER`. This env can be used to get privileged access by other methods than `resvport`. The `gss` method in this case. This option opens a possibility to disable `resvport` and use only `gss` method for all interaction with OpenPBS in case of such a desire. 

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->

1. Added support to subnets in `acl_hosts` checks IP address obtained by a forward resolution of a request's hostname to the provided `subnet/mask`. The `subnet/mask` is distinguished by `inet_pton` and requires `/` char. Hence it can be safely added to `acl_hosts`.

2.  The `acl_hosts` are not checked for the request `PBS_BATCH_Connect` now. The reason for this is that `PBS_BATCH_Authenticate` is sent after `PBS_BATCH_Connect` and at the time of `PBS_BATCH_Connect` we can not know which auth method will the connection use. Using the `gss` auth method, the `acl_hosts` were used for the `PBS_BATCH_Connect` even if the connection is authenticated in `PBS_BATCH_Authenticate` and can safely use the `acl_krb_realms`.  This change is safe because no other request can be read before `PBS_BATCH_Authenticate` except for simple `PBS_BATCH_Connect`. An attempt to send a different request before `PBS_BATCH_Authenticate` will result in `PBSE_BADCRED`.

3. A new server env `PBS_PRIVILEGED_AUTH_USER` is added. Privileged permission is checked against user and host in each request. Using `gss`/kerberos, the user and host contain (and together create) a principal. Using `gss`/Kerberos, the new env expects principal and is simply used as a way how to get privileged permissions. 

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->


#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->

* PTL results are attached: [ptl_acl_subnets.txt](https://github.com/openpbs/openpbs/files/14370552/ptl_acl_subnets.txt)

Manual test of privileged access, using only gss:
```
(BULLSEYE)root@torque4:~# grep AUTH /etc/pbs.conf 
PBS_AUTH_METHOD=GSS
PBS_SUPPORTED_AUTH_METHODS=GSS
PBS_DAEMON_SERVICE_AUTH_USER=host/torque4.grid.cesnet.cz@EINFRA-SERVICES
PBS_PRIVILEGED_AUTH_USER=host/torque4.grid.cesnet.cz@EINFRA-SERVICES
```
the privileged access is granted to the selected principal:
```
(BULLSEYE)root@torque4:~# KRB5CCNAME=/tmp/krb5cc_pbs_client klist
Credentials cache: FILE:/tmp/krb5cc_pbs_client
        Principal: host/torque4.grid.cesnet.cz@EINFRA-SERVICES

  Issued                Expires               Principal
Feb 21 04:44:18 2024  Feb 22 04:44:18 2024  krbtgt/EINFRA-SERVICES@EINFRA-SERVICES
Feb 21 04:44:18 2024  Feb 22 04:44:18 2024  host/torque4.grid.cesnet.cz@EINFRA-SERVICES
(BULLSEYE)root@torque4:~# qmgr -c 'l h'
Hook test
    type = site
    enabled = true
    event = ""
    user = pbsadmin
    alarm = 30
    order = 1
    debug = false
    fail_action = none

(BULLSEYE)root@torque4:~#
```
and the access is rejected for others:
```
(BULLSEYE)root@torque5:~#  KRB5CCNAME=/tmp/krb5cc_pbs_client klist
Credentials cache: FILE:/tmp/krb5cc_pbs_client
        Principal: host/torque5.grid.cesnet.cz@EINFRA-SERVICES

  Issued                Expires               Principal
Feb 21 12:00:28 2024  Feb 22 12:00:28 2024  krbtgt/EINFRA-SERVICES@EINFRA-SERVICES
Feb 21 12:00:28 2024  Feb 22 12:00:28 2024  host/torque4.grid.cesnet.cz@EINFRA-SERVICES
(BULLSEYE)root@torque5:~# qmgr -c 'l h'
host/torque5.grid.cesnet.cz@EINFRA-SERVICES is unauthorized to access hooks data from server torque4.grid.cesnet.cz
(BULLSEYE)root@torque5:~# 

```
also  resvport is rejected:
```
(BULLSEYE)root@torque4:~# PBS_AUTH_METHOD=resvport PBS_ENCRYPT_METHOD= qmgr
auth: error returned: -1
auth: Unable to authenticate connection (torque4.grid.cesnet.cz:15001)
Invalid argument
qmgr: cannot connect to server 
(BULLSEYE)root@torque4:~# 

```

<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

SudeshnaMoulik commented:
Regarding the changes made for option 2:

We don't want to allow a user to connect to server, if its not allowed in acl_host (if that attribute is enabled). 
But with the changes via this MR, now even though user is not allowed to connect to server, they will be able to do so, as the acl verification is moved to PBS_BATCH_Authenticate request. 

I think this will impose security risk, as user can send any malicious request to server (Or probably then can even make the server hang) as they are already connected to the server.

In my opinion, we should let the acl_host verification happen while connecting to PBS Server in PBS_BATCH_Connect and let acl_krb_realms verify for further granular verification in realm level in PBS_BATCH_Authenticate. It would act as safety guard irrespective of what auth method set in config. What do you think ?

SudeshnaMoulik commented:
What are the allowed format for PBS_PRIVILEGED_AUTH_USER in config ?

SudeshnaMoulik commented on src/server/process_request.c:
```diff
@@ -596,7 +597,9 @@ process_request(int sfds)
 
 	/* is the request from a host acceptable to the server */
-	if (access_by_krb == 0 && get_sattr_long(SVR_ATR_acl_host_enable)) {
+	if (request->rq_type != PBS_BATCH_Connect &&
```
If Kerberos is enabled, then this will not get verified again for any any request type. So how about not mentioning "request->rq_type != PBS_BATCH_Connect " ?

SudeshnaMoulik commented on src/server/process_request.c:
```diff
@@ -569,9 +570,9 @@ process_request(int sfds)
-	/* FIXME: Do we need realm check for all auth ? */
 #if defined(PBS_SECURITY) && (PBS_SECURITY == KRB5)
-	if (conn->cn_credid != NULL &&
+	if (request->rq_type != PBS_BATCH_Connect &&
```
Why we don't want it to verify for PBS_BATCH_Connect ?
Because it verifies whether server attribute acl_krb_realm_enable set.

vchlum replied:
Removed. This was useless anyway. The `conn->cn_credid` and other `conn` attributes are not set in the `PBS_BATCH_Connect`. Will be set after authentication.

SudeshnaMoulik commented on src/server/process_request.c:
```diff
@@ -608,9 +611,33 @@ process_request(int sfds)
+
+			access_allowed = 0;
+
+			if (acl_check(get_sattr(SVR_ATR_acl_hosts), request->rq_host, ACL_Host)) {
```
Can we make this acl subnet verification in a separate function and use it irrespective of access_by_krb ?

vchlum replied:
OK. Since we removed the condition `request->rq_type != PBS_BATCH_Connect` above, the variable `access_by_krb` has no meaning anymore. I removed the variable. I also changed the order of evaluation `acl_hosts` and `acl_krb_realms`. `acl_hosts` is first now. It makes more sense. Also, the special case when we access the `server` from the `server` itself is considered here.

SudeshnaMoulik commented on test/tests/functional/pbs_acl_subnets.py:
```diff
@@ -0,0 +1,121 @@
+from tests.functional import *
+
+
+class Test_acl_subnet(TestFunctional):
```
As we are enhancing "acl_hosts" server feature to allow subnet, I would recommend to use the test suite name as "Test_acl_host_server"

vchlum replied:
OK

vchlum commented on src/include/pbs_ifl.h:
```diff
@@ -486,6 +486,7 @@ enum mgr_obj {
 #define PBS_STDNG_RESV_ID_CHAR 'S'						     /* Character in front of a resv ID */
 #define PBS_MNTNC_RESV_ID_CHAR 'M'						     /* Character in front of a resv ID */
 #define PBS_AUTH_KEY_LEN (129)
+#define PBS_MAXIP 15
```
It is the length of an IP address. I made the name more obvious and added a comment.

vchlum commented on src/include/pbs_internal.h:
```diff
@@ -312,6 +313,7 @@ extern struct pbs_config pbs_conf;
 #define PBS_CONF_SCHED_THREADS	"PBS_SCHED_THREADS"
 #define PBS_CONF_DAEMON_SERVICE_USER "PBS_DAEMON_SERVICE_USER"
 #define PBS_CONF_DAEMON_SERVICE_AUTH_USER "PBS_DAEMON_SERVICE_AUTH_USER"
+#define PBS_CONF_PRIVILEGED_AUTH_USER "PBS_PRIVILEGED_AUTH_USER"
```
I added a check whether the GSS is between PBS_SUPPORTED_AUTH_METHODS. Make sense to check it. It is currently used only if pbs is compilled with gss/krb. So it expects a Kerberos principal - comment added.

vchlum commented:
Thank you for the comments. I addressed the comments @SudeshnaMoulik. I think ignoring the acl_hosts for PBS_BATCH_Connect does not mean security risk unless we add some read between `PBS_BATCH_Connect` and `PBS_BATCH_Authenticate` or even add some reading in `PBS_BATCH_Connect`. I understand your point. I also did not consider the possibility of hanging the server. Thank you.

The PBS_PRIVILEGED_AUTH_USER is expected to be a host principal, like `host/<fqdn>@<REALM>`. The format is not forced but it is set by the admins and it is compared only to principals.

I would like to share one more idea, please let me know if it is worth following: How about adding new server attributes like `acl_hosts_gss` and `acl_hosts_iff`? These attributes would be evaluated independently on `acl_hosts` after the `PBS_BATCH_Authenticate` based on the auth method. In a firewall-safe environment, it would allow disabling `acl_hosts` and use only the new attributes.

SudeshnaMoulik commented on src/include/pbs_internal.h:
```diff
@@ -313,7 +313,7 @@ extern struct pbs_config pbs_conf;
 #define PBS_CONF_DAEMON_SERVICE_USER "PBS_DAEMON_SERVICE_USER"
 #define PBS_CONF_DAEMON_SERVICE_AUTH_USER "PBS_DAEMON_SERVICE_AUTH_USER"
-#define PBS_CONF_PRIVILEGED_AUTH_USER "PBS_PRIVILEGED_AUTH_USER"
+#define PBS_CONF_PRIVILEGED_AUTH_USER "PBS_PRIVILEGED_AUTH_USER" /* e.g.: used for gss/krb and krb principal is expected */
```
Please add an example of the allowed format

vchlum replied:
done

SudeshnaMoulik commented on src/server/process_request.c:
```diff
@@ -645,6 +615,41 @@ process_request(int sfds)
+			access_allowed = 1;
+		}
+
+		if (acl_check(get_sattr(SVR_ATR_acl_krb_realms), conn->cn_credid, ACL_Host)) {
```
Before this check, please check whether its enabled "if (get_sattr_long(SVR_ATR_acl_krb_realm_enable))"

vchlum replied:
Thx, done

SudeshnaMoulik commented:
> Thank you for the comments. I addressed the comments @SudeshnaMoulik. I think ignoring the acl_hosts for PBS_BATCH_Connect does not mean security risk unless we add some read between `PBS_BATCH_Connect` and `PBS_BATCH_Authenticate` or even add some reading in `PBS_BATCH_Connect`. I understand your point. I also did not consider the possibility of hanging the server. Thank you. I removed critical condition.
> 
> The PBS_PRIVILEGED_AUTH_USER is expected to be a host principal, like `host/<fqdn>@<REALM>`. The format is not forced but it is set by the admins and it is compared only to principals.
> 
> I would like to share one more idea, please let me know if it is worth following: How about adding new server attributes like `acl_hosts_gss` and `acl_hosts_iff`? These attributes would be evaluated independently on `acl_hosts` after the `PBS_BATCH_Authenticate` based on the auth method. In a firewall-safe environment, it would allow disabling `acl_hosts` and use only the new attributes.

I think the enhancement of "acl_host_ to accepts subnets too, is generic and not specific to GSS. Moreover "`acl_hosts_gss` and `acl_hosts_iff`" will also be allowed hosts list, so its better to handle them only with acl_hosts which accepts subnet.

SudeshnaMoulik reviewed (approved):
Thanks @vchlum for addressing all the comments 👍

bayucan reviewed (approved):
Looks good.
